### ISSUE TITLE
CQLTranslator doesn't escape special characters for String values

### ISSUE THREAD

**Issue by dtuzikov** [Impetus/kundera#230]
Hi.

Method CQLTranslator.appendValue doesn't escape special characters for String values. For example quote symbol ' should be escaped with extra ' in CQL

Kundera 2.4

**Comment by mevivs** [comment]
Will look into this.

-Vivek

**Comment by mevivs** [comment]
Fixed. Please verify with 2.5 release

-Vivek

**Comment by dtuzikov** [comment]
Not completely fixed.

``` java
@Embeddable
public class UserDataPK implements Serializable {
    @Column(name = "user_key")
    private String userKey;

    @Column(name = "user_sys")
    private String userSys;

    //... constructor, getters, setters 
}

@Entity
@Table(name = "user_data", schema = "sch001@cassandra_pu")
public class UserData {
    @EmbeddedId
    protected UserDataPK pk;

    @Column (name = "some_data")
    private String someData;

    //... constructor, getters, setters 
}

public class Test  {
    @PersistenceContext(type = PersistenceContextType.EXTENDED)
    private EntityManager em;

    @Test
    public void test1() {
        final String value="a_'_b";
        final String quoteWrappedValue="'"+value+"'";

        UserData userData=new UserData();
        userData.setPk(new UserDataPK(quoteWrappedValue,"SYS_X"));
        userData.setSomeData(quoteWrappedValue);
        em.persist(userData);
        //refresh to see what is really stored to cassandra
        em.refresh(userData);
        //reread
        userData=em.find(UserData.class,new UserDataPK(quoteWrappedValue,"SYS_X"));
        assertNotNull(userData);

        assertEquals(userData.getSomeData(),value); //<-- OK
        assertEquals(userData.getPk().getUserKey(),value); //<-- OK

        assertEquals(userData.getSomeData(),quoteWrappedValue); //<-- FAILS
        assertEquals(userData.getPk().getUserKey(),quoteWrappedValue); //<-- FAILS
    }
}
```

**Comment by mevivs** [comment]
I guess it's incorrect way of defining:

```
 final String value="a_'_b";
        final String quoteWrappedValue="'"+value+"'";
```

As per JPA:

```
A string literal is enclosed in single quotes—for example: ‘literal’. A string literal that includes a single
quote is represented by two single quotes—for example: ‘literal’’s’. String literals in queries, like Java
String literals, use unicode character encoding. The use of Java escape notation is not supported in
query string literals.

```

So given below should work:

```
 final String value="a_''_b";
        final String quoteWrappedValue="'"+value+"'";
```

-Vivek

**Comment by dtuzikov** [comment]
The error comes with left and right quotation marks, quote between ab persists normally.

**Comment by dtuzikov** [comment]
The problem is in appendValue() method of CQLTranslator. It removes first and last quotes from string values before escaping quote. 

``` java
//...
   if (fieldClazz.isAssignableFrom(String.class))
   {
      // To allow escape character
      value = ((String) value).replaceAll("^'", "").replaceAll("'$", "").replaceAll("'", "''");
   }
//...
```

**Comment by mevivs** [comment]
So, if i am getting it correctly, as per your test case you want to persist 'a_'b' rather than a_'b ? 

Did you try same persisting via CQLsh as a cql query?

-Vivek

**Comment by dtuzikov** [comment]
Yes, i want to persist 'a_'b', so CQLTranslator should escape all quotes like ''a_''b''.
cql3 query: 

```
insert into user_data (user_key, user_sys, some_data) 
       values ('''a_''_b''','SYS_X','''a_''_b'''); 
```

works well from cqlsh:

```
cqlsh>select * from user_data;
     user_key | user_sys | some_data 
--------------+----------+--------------
      'a_'_b' |   SYS_X  |  'a_'_b'
```

**Comment by mevivs** [comment]
Right. So as per your CQL query to persist 'a_'b', you explicitly enclose it as  

```
'''a_''_b'''
```

With Kundera, everything such queries would be converted into CQL queries, and Kundera simply strip enclosed "'". as per JPA rules. 
So if you define:

```
 final String value="a_'_b";
        final String quoteWrappedValue="'"+value+"'";
```

AS

`
 final String value="a_'_b";
        final String quoteWrappedValue="''"+value+"''";
`

It should work.

However, i understand that such translations can be implicitly handled within Kundera while translating queries.

-Vivek

**Comment by dtuzikov** [comment]
I think escaping code should be changed from

``` java
//...
   if (fieldClazz.isAssignableFrom(String.class))
   {
      // To allow escape character
      value = ((String) value).replaceAll("^'", "").replaceAll("'$", "").replaceAll("'", "''");
   }
//...
```

to

``` java
//...
   if (fieldClazz.isAssignableFrom(String.class))
   {
      // To allow escape character
      value = ((String) value).replaceAll("'", "''");
   }
//...
```

**Comment by mevivs** [comment]
Fixed and releasing with 2.6. Please verify.
-Vivek

**Comment by dtuzikov** [comment]
fixed
